ops/charm.py

```diff
@@ -147,6 +147,162 @@ class HookEvent(EventBase):
     """
 
 
+class SecretEvent(HookEvent):
+    """Base class for secret events."""
+
+    # used to determine whether the secret instance
+    # exposed by this event is self-owned or not. Uses the knowledge that certain event
+    # types are only fired on secret owners / consumers.
+    _is_owner_event = NotImplemented
+
+    if TYPE_CHECKING:
+        _SecretEventSnapshot = TypedDict('_SecretEventSnapshot', {
+            'id': str,
+            'label': str,
+            'revision': int,
+        })
+
+    def __init__(self,
+                 handle: 'Handle',
+                 id: str,
+                 label: Optional[str] = None,
+                 revision: Optional[int] = None):
+        super().__init__(handle)
+        assert isinstance(self._is_owner_event, bool), '_is_owner_event ' \
+                                                       'should be overridden in subclass'
+        self.id = id
+        self.label = label
+        self.revision = revision
+
+    @property
+    def secret(self) -> model.Secret:  # pyright: reportPrivateUsage=false
+        """The secret instance this event is about."""
+        return model.Secret(  # pyright: reportPrivateUsage=false
+            backend=self.framework.model._backend,  # pyright: reportPrivateUsage=false
+            id=self.id,
+            label=self.label,
+            revision=self.revision,
+            is_owned_by_this_unit=self._is_owner_event)
+
+    def snapshot(self) -> '_SecretEventSnapshot':
+        """Used by the framework to serialize the event to disk.
+
+        Not meant to be called by charm code.
+        """
+        # relax the _SerializedData type a bit:
+        snapshot = cast(Dict[str, Optional['JsonObject']], super().snapshot())
+        snapshot['id'] = self.id
+        snapshot['label'] = self.label
+        snapshot['revision'] = self.revision
+        return cast('SecretEvent._SecretEventSnapshot', snapshot)
+
+    def restore(self, snapshot: '_SecretEventSnapshot'):
+        """Used by the framework to deserialize the event from disk.
+
+        Not meant to be called by charm code.
+        """
+        super().restore(snapshot)  # type: ignore
+        self.id = snapshot['id']
+        self.label = snapshot['label']
+        self.revision = snapshot['revision']
+
+
+class SecretChangedEvent(SecretEvent):
+    """Event raised by Juju on a secret **consumer** when the owner has changed its contents.
+
+    When the owner of a secret creates a new revision, all units/applications
+    that are tracking this secret will be notified by means of this event that a new
+    revision is available.
+
+    Typically, you will want to update the secret (i.e. fetch the new revision)
+    and prune the old one.
+
+    Attributes:
+        secret: class:`Secret`, the Secret instance this event is about.
+
+    >>> def _on_secret_changed(self: CharmBase, evt: SecretChangedEvent):
+    >>>     evt.secret.get_latest_revision()
+    >>>     evt.secret.prune()
```

Comment:
I'm confused about prune. It seems we intend prune to be used by the owner to remove obsolete revisions. So why include it here when talking about a consumer responding to a secret change event?